localstack-core/localstack/aws/connect.py

```diff
@@ -631,9 +652,88 @@ def get_client(
         )
 
 
+def resolve_dns_from_upstream(hostname: str) -> str:
+    upstream_dns = get_fallback_dns_server()
+    request = dns.message.make_query(hostname, "A")
+    response = dns.query.udp(request, upstream_dns, port=53, timeout=5)
+    if len(response.answer) == 0:
+        raise ValueError(f"No DNS response found for hostname '{hostname}'")
+
+    ip_addresses = list(response.answer[0].items.keys())
+    return choice(ip_addresses).address
+
+
+class ExternalBypassDnsClientFactory(ExternalAwsClientFactory):
+    """
+    Client factory that makes requests against AWS ensuring that DNS resolution is not affected by the LocalStack DNS
+    server.
+    """
+
+    def __init__(
+        self,
+        session: Session = None,
+        config: Config = None,
+    ):
+        super().__init__(use_ssl=True, verify=True, session=session, config=config)
+
+    def _get_client_post_hook(self, client: BaseClient) -> BaseClient:
+        client = super()._get_client_post_hook(client)
+        client._endpoint.http_session = ExternalBypassDnsSession()
+        return client
+
+
+class ExternalBypassDnsHTTPConnection(AWSHTTPConnection):
+    """
+    Connection class that bypasses the LocalStack DNS server for HTTP connections
+    """
+
+    def _new_conn(self) -> socket:
+        orig_host = self._dns_host
+        try:
+            self._dns_host = resolve_dns_from_upstream(self._dns_host)
+            return super()._new_conn()
+        finally:
+            self._dns_host = orig_host
+
+
+class ExternalBypassDnsHTTPSConnection(AWSHTTPSConnection):
+    """
+    Connection class that bypasses the LocalStack DNS server for HTTPS connections
+    """
+
+    def _new_conn(self) -> socket:
+        orig_host = self._dns_host
+        try:
+            self._dns_host = resolve_dns_from_upstream(self._dns_host)
+            return super()._new_conn()
+        finally:
+            self._dns_host = orig_host
+
+
+class ExternalBypassDnsHTTPConnectionPool(AWSHTTPConnectionPool):
+    ConnectionCls = ExternalBypassDnsHTTPConnection
+
+
+class ExternalBypassDnsHTTPSConnectionPool(AWSHTTPSConnectionPool):
+    ConnectionCls = ExternalBypassDnsHTTPSConnection
+
+
+class ExternalBypassDnsSession(URLLib3Session):
```

Comment:
nit: Is there a chance to see the new session / client being tested here somehow? While this approach seems to be more stable, it still relies on some internals (like `_pool_classes_by_scheme`). Would be great to have a test which immediately informs us if this client won't work anymore (f.e. directly in the ASF update PRs where we bump the `botocore` version weekly).

Reply:
I have created a follow up ticket to make sure it doesn't fall off our radar. I should be good to tackle it this week. I will merge for now to help unblock further progress!